hudi-utilities/src/main/java/org/apache/hudi/utilities/HoodieMetadataTableValidator.java

```diff
@@ -1115,6 +1145,86 @@ private void validateSecondaryIndex(HoodieSparkEngineContext engineContext, Hood
+    Dataset<Row> mdtIndexDf = functionalIndexSupport.loadFunctionalIndexDataFrame(indexPartition, false).cache();
+    Dataset<Row> fsFunctionalIndexDf = getFSFunctionalIndexRecords(engineContext, metaClient.getBasePath(), indexDefinition, metadataContext)
+        .select(Arrays.stream(FunctionalIndexSupport.getTargetColumnStatsIndexColumns()).map(functions::col).toArray(Column[]::new))
+        .cache();
+    Dataset<Row> fsFunctionalIndexDiff = fsFunctionalIndexDf.exceptAll(mdtIndexDf);
+    List<String> diffRecordsFilenames = fsFunctionalIndexDiff.select(HoodieMetadataPayload.COLUMN_STATS_FIELD_FILE_NAME)
+        .map((MapFunction<Row, String>) row -> row.getString(0), Encoders.STRING())
+        .collectAsList();
+    Dataset<Row> mdtFunctionalIndexDiffRecords = mdtIndexDf.filter((FilterFunction<Row>) row -> {
+      String fileName = row.getAs(HoodieMetadataPayload.COLUMN_STATS_FIELD_FILE_NAME);
+      return diffRecordsFilenames.contains(fileName);
+    });
+    if (diffRecordsFilenames.isEmpty()) {
+      throw new HoodieValidationException(String.format("Functional Index does not match : \nFS functional index diff records: %s \nCorresponding MDT functional index diff records: %s",
+          Arrays.toString(fsFunctionalIndexDf.collectAsList().toArray()), Arrays.toString(mdtFunctionalIndexDiffRecords.collectAsList().toArray())));
+    }
+    fsFunctionalIndexDf.unpersist();
+    mdtIndexDf.unpersist();
+  }
+
+  Dataset<Row> getFSFunctionalIndexRecords(HoodieSparkEngineContext sparkEngineContext, StoragePath basePath, HoodieIndexDefinition indexDefinition,
+                                           HoodieMetadataValidationContext metadataContext) throws Exception {
+    // Get latest partition file slices and generate functional index records using functional index initialization API
+    List<Pair<String, Pair<String, Long>>> partitionFilePathSizeTriplet = getPartitionFileSlicePairs(metaClient, metadataContext.tableMetadata, metadataContext.fileSystemView);
+    Schema readerSchema = getProjectedSchemaForFunctionalIndex(indexDefinition, metaClient);
+    StructType columnStatsRecordStructType = AvroConversionUtils.convertAvroSchemaToStructType(HoodieMetadataColumnStats.SCHEMA$);
+    HoodieWriteConfig writeConfig = HoodieWriteConfig.newBuilder().withPath(basePath).build();
+    HoodieData<InternalRow> fsFunctionalIndexRecords = getFunctionalIndexRecords(partitionFilePathSizeTriplet, indexDefinition, metaClient, 10, readerSchema,
+        metadataContext.getMetaClient().getStorageConf(), "", sparkEngineContext, writeConfig, writeConfig)
+        .map(record -> {
+          HoodieMetadataPayload metadataPayload = (HoodieMetadataPayload) record.getData();
+          return AvroConversionUtils.createAvroToInternalRowConverter(HoodieMetadataColumnStats.SCHEMA$, columnStatsRecordStructType).apply(metadataPayload.getColumnStatMetadata().get()).orNull(null);
+        });
```

Comment:
Why do we need to convert here? Can we not have a FunctionalIndexSupport API to return metadata payload itself instead of Dataset<Row> because that support class anyway uses the metadata table reader which returns the payload and then we convert to dataframe.

Reply:
This is actually beneficial since the schema of the dataset is only reflecting column stat fields. It makes it easier to compare and query.